core/dbt/task/test.py

```diff
@@ -142,8 +144,12 @@ def execute_data_test(self, data_test: TestNode, manifest: Manifest) -> TestResu
 
         # generate materialization macro
         macro_func = MacroGenerator(materialization_macro, context)
-        # execute materialization macro
-        macro_func()
+        try:
+            # execute materialization macro
+            macro_func()
+        finally:
+            self.adapter.post_model_hook(context, hook_ctx)
```

Comment:
To confirm, this is not a behaviour change because: 
* pre/post_model_hook are no-ops by default: https://github.com/dbt-labs/dbt-adapters/blob/8ff69435a753f0a94dd1274d0e8dad199805cd13/dbt/adapters/base/impl.py#L1415-L1438
* only [dbt-snowflake has custom implementation](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/adapters/snowflake/impl.py#L101), which is based on the presence of a config and effectively a no-op otherwise (uses warehouse from profile). Is the assumption that users with a `snowflake_warehouse` would _always_ want to route tests to that warehouse? 

cc @jtcohen6

Reply:
I think i can help supply a answer here. with this pr and 

- https://github.com/dbt-labs/dbt-core/pull/10245 will allow us to configure the snowflake_warehosue even for tests as its a config that can be changed in various ways through the config:

Reply:
ah! gotcha -- so configs were not possible to provide for generic tests in the past, which means there should not be any currently set on them

Reply:
from my understanding yes, we could use a old pattern to set a set number of accepted configs linked in the generic_config issue, making this update to the newer syntax while maintaining the old will allow us to do this now.

Reply:
Thanks @MichelleArk @McKnight-42 for thinking of this!

As I see it, the only way in which this could be a "breaking" change is if someone had configured a singular test, or all tests (from `dbt_project.yml`), with a `snowflake_warehouse` that they didn't actually want to be used.

Something like:
```
# dbt_project.yml
tests:
  +snowflake_warehouse: does_not_exist
```
```
-- tests/my_singular_test.sql
{{ config(snowflake_warehouse = 'wrong_warehouse') }}

select ...
```

I do not see that as a breaking change; I see it as this feature in dbt (specifically `dbt-snowflake`) finally working as intended. As such, I do not believe we need to put it behind a legacy behavior migration flag.